tls/s2n_ktls_io.c

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_ktls.h"
+#include "utils/s2n_socket.h"
+
+/* Used to override sendmsg and recvmsg for testing. */
+static ssize_t s2n_ktls_default_sendmsg(void *io_context, const struct msghdr *msg);
+static ssize_t s2n_ktls_default_recvmsg(void *io_context, struct msghdr *msg);
+s2n_ktls_sendmsg_fn s2n_sendmsg_fn = s2n_ktls_default_sendmsg;
+s2n_ktls_recvmsg_fn s2n_recvmsg_fn = s2n_ktls_default_recvmsg;
+
+S2N_RESULT s2n_ktls_set_sendmsg_cb(struct s2n_connection *conn, s2n_ktls_sendmsg_fn send_cb, void *send_ctx)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(send_ctx);
+    RESULT_ENSURE(s2n_in_test(), S2N_ERR_NOT_IN_TEST);
+    conn->send_io_context = send_ctx;
+    s2n_sendmsg_fn = send_cb;
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_set_recvmsg_cb(struct s2n_connection *conn, s2n_ktls_recvmsg_fn recv_cb, void *recv_ctx)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(recv_ctx);
+    RESULT_ENSURE(s2n_in_test(), S2N_ERR_NOT_IN_TEST);
+    conn->recv_io_context = recv_ctx;
+    s2n_recvmsg_fn = recv_cb;
+    return S2N_RESULT_OK;
+}
+
+static ssize_t s2n_ktls_default_recvmsg(void *io_context, struct msghdr *msg)
+{
+    POSIX_ENSURE_REF(io_context);
+    POSIX_ENSURE_REF(msg);
+
+    const struct s2n_socket_read_io_context *peer_socket_ctx = io_context;
+    int fd = peer_socket_ctx->fd;
```

Comment:
https://github.com/aws/s2n-tls/pull/4109#discussion_r1286624115 You need to check the context before dereferencing it. This is even more important because this is the non-test path.
```suggestion
    const struct s2n_socket_read_io_context *peer_socket_ctx = io_context;
    POSIX_ENSURE_REF(peer_socket_ctx);
    int fd = peer_socket_ctx->fd;
```